utils/env_preferences/src/parse/posix.rs

```diff
@@ -0,0 +1,621 @@
+            Some((offset, _delimiter)) => &src[..*offset],
+            None => src,
+        };
+
+        // Make sure the language itself is non-empty
+        if language.is_empty() {
+            return Err(ParseError::EmptySection { offset: 0 });
+        }
+
+        let mut locale = Self {
+            language,
+            territory: None,
+            codeset: None,
+            modifier: None,
+        };
+
+        for (index, (start_offset, delimiter)) in optional_sections.iter().enumerate() {
+            // Find the offset of the next section, or end of the string if none exist
+            let end_offset = optional_sections
+                .get(index + 1)
+                .map(|(next_offset, _next_delimiter)| *next_offset)
+                .unwrap_or(src.len());
+
+            // Make sure this section is non-empty (more characters than just the delimiter)
+            if start_offset + 1 >= end_offset {
+                return Err(ParseError::EmptySection {
+                    offset: *start_offset,
+                });
+            }
+
+            // Write the section to the appropriate field
+            let section_value = Some(&src[start_offset + 1..end_offset]);
+            match delimiter {
+                Delimiter::Territory => locale.territory = section_value,
+                Delimiter::Codeset => locale.codeset = section_value,
+                Delimiter::Modifier => locale.modifier = section_value,
+            }
+        }
+
+        Ok(locale)
+    }
+
+    /// Attempt to convert a POSIX locale into a valid BCP-47 identifier.
+    ///
+    /// This is a best-effort conversion process, and there are valid
+    /// POSIX locales that will return an error or silently ignore data.
+    /// In particular, the codeset section is always ignored, and only some common modifiers are handled
+    /// (unknown modifiers will be silently ignored).
+    ///
+    /// If the "logging" feature is enabled, this ignored data will be logged as a warning.
+    pub fn try_convert_lossy(&self) -> Result<Locale, ConversionError> {
+        // Check if the language needs to be mapped to a BCP-47 language subtag
+        let language_code = POSIX_LANGUAGE_NAMES
+            .binary_search_by_key(&self.language, |(language, _code)| language)
+            // `binary_search_by_key()` guarantees a valid index
+            .map(|index| POSIX_LANGUAGE_NAMES[index].1);
+
+        let mut language = match language_code {
+            // Found a matching language code
+            Ok(language) => language,
+            // Parse the language directly
+            Err(_) => match Language::try_from_str(self.language) {
+                Ok(language) => language,
+                Err(_) => {
+                    return Err(ConversionError::InvalidLanguage {
+                        start_offset: 0,
+                        end_offset: self.language.len(),
+                    })
+                }
+            },
+        };
```

Comment:
`nit`: you can make this code a bit cleaner by using chaining:

```rust
let language = get_bcp47_language_from_posix(&self.language)
    .or(|| Language::try_from_str(self.language).ok())
    .ok_or(ConversionError::InvalidLanguage { start_offset: 0, end_offset: self.language.len()})?;
```

Reply:
Much cleaner, thank you :)